streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+    all_aud_playlists: List[List['MediaPlaylist']] = []
+    all_var_playlists: List[List['MediaPlaylist']] = []
+    # The durations will be used to insert a gap in subtitles playlists when
+    # there is no applicable media to insert. Will also be used to calculate
+    # the average bandwidth.
+    durations: List[float] = []
+    
+    for master_playlist in master_playlists:
+      
+      txt_playlists: List['MediaPlaylist'] = []
+      aud_playlists: List['MediaPlaylist'] = []
+      var_playlists: List['MediaPlaylist'] = []
+      
+      for media_playlist in master_playlist.playlists:
+        stream_type = media_playlist.stream_info.get('TYPE', 'STREAM-INF')
+        if stream_type == 'SUBTITLES':
+          txt_playlists.append(media_playlist)
+        elif stream_type == 'AUDIO':
+          aud_playlists.append(media_playlist)
+        elif stream_type == 'STREAM-INF':
+          var_playlists.append(media_playlist)
+        else:
+          # TODO: We need a case for CLOSED-CAPTIONS(CC).
+          raise RuntimeError("TYPE={} is not regonized.".format(stream_type))  
+      
+      all_txt_playlists.append(txt_playlists)
+      all_aud_playlists.append(aud_playlists)
+      all_var_playlists.append(var_playlists)
+      durations.append(master_playlist.duration)
+    
+    master_hls = MasterPlaylist()
+    master_hls.playlists.extend(
+        MediaPlaylist.concat_sub(all_txt_playlists, durations))
+    
+    if all(not MediaPlaylist.inf_is_vid(inf_pl) for inf_pl in all_var_playlists):
+      # When the playlist is audio only, each audio is referenced two times,
+      # once in an #EXT-X-MEDIA tag and another time in an #EXT-X-STREAM-INF tag.
+      # If the user has an audio-only content, the concatenation will go a little
+      # bit different to produce the desired output.
+      master_hls.playlists.extend(
+          MediaPlaylist.concat_aud_only(
+            all_aud_playlists,
+            all_var_playlists,
+            durations))
+    else:
+      master_hls.playlists.extend(
+          MediaPlaylist.concat_aud(all_aud_playlists))
+      master_hls.playlists.extend(
+          MediaPlaylist.concat_vid(all_var_playlists, durations))
+    
+    return master_hls
+
+
+class HLSConcater:
+  """A class that serves as an API for the m3u8 concatenation methods."""
+  
+  def __init__(self,
+               sample_master_playlist_path: str,
+               output_location: str):
+    """Calls MasterPlyalist.extract_headers() and MediaPlaylist.extract_header()
+    and store these headers in their classes respectively.
+    
+    We use the `output_location` to re-evaluate a media segment paths, and also write
+    the concatenation result to this `output_location`.
+    """
+    
+    # Save common master playlist header, this will call
+    # MediaPlaylist.extract_header() to save the common
+    # media playlist header as well.
+    MasterPlaylist.extract_headers(sample_master_playlist_path)
```

Comment:
It bothers me that these are stored statically on the class.  It means that an application using the Python API can't create multiple independent controllers doing concatenation if the headers would need to be different.

What if, instead, the headers (both master & media headers) are stored on the HLSConcater instance, and passed to .write()?